src/lib/spinel/radio_spinel.cpp

```diff
@@ -2457,5 +2460,73 @@ void RadioSpinel::HandleCompatibilityError(void)
     DieNow(OT_EXIT_RADIO_SPINEL_INCOMPATIBLE);
 }
 
+void RadioSpinel::Metrics::Init() { ReadMetrics(*reinterpret_cast<otRadioSpinelMetrics *>(this)); }
+
+const otRadioSpinelMetrics *RadioSpinel::Metrics::GetMetrics(void) const
+{
+    return reinterpret_cast<const otRadioSpinelMetrics *>(this);
+}
+
+bool RadioSpinel::Metrics::IsReadWriteCallbackValid(void) const
+{
+    return (mCallbacks.mWriteRadioSpinelMetrics != nullptr) && (mCallbacks.mReadRadioSpinelMetrics != nullptr);
+}
+
+void RadioSpinel::Metrics::ReadMetrics(otRadioSpinelMetrics &aMetrics)
+{
+    VerifyOrExit(IsReadWriteCallbackValid());
+
+    if (mCallbacks.mReadRadioSpinelMetrics(aMetrics, mCallbacks.mRadioSpinelMetricsContext) != OT_ERROR_NONE)
+    {
+        memset(&aMetrics, 0, sizeof(aMetrics));
+        mCallbacks.mWriteRadioSpinelMetrics(aMetrics, mCallbacks.mRadioSpinelMetricsContext);
+    }
+
+exit:
+    return;
+}
+
+void RadioSpinel::Metrics::WriteMetrics(otRadioSpinelMetrics &aMetrics)
+{
+    VerifyOrExit(IsReadWriteCallbackValid());
+    mCallbacks.mWriteRadioSpinelMetrics(aMetrics, mCallbacks.mRadioSpinelMetricsContext);
+
+exit:
+    return;
+}
```

Comment:
The `ReadMetrics` and `WriteMetrics` methods have an input `otRadioSpinelMetrics &aMetrics`, but they are always called with `*this`. I suggest making it simpler and having them directly use the object itself?